File: 2020/02.js
const fs = require("fs");
fs.readFile("inputs/02.in", "utf8", main);

function main(err, data) {
  const rows = data.split("\n").slice(0, -1).map(parseRow);
  console.log("Part1:", rows.map(part1).reduce((x, y) => x + y));
  console.log("Part2:", rows.map(part2).reduce((x, y) => x + y));
}

function parseRow(row) {
  const parts = row.split(" ");
  const range = parts[0].split("-");
  return {
    low: Number(range[0]),
    high: Number(range[1]),
    policy: parts[1].slice(0, -1),
    password: parts[2],
  };
}

function part1(row) {
  const polCount = [...row.password].filter((x) => x === row.policy).length;
  return polCount >= row.low && polCount <= row.high;
}

function part2(row) {
  const check1 = row.password[row.low - 1] === row.policy;
  const check2 = row.password[row.high - 1] === row.policy;
  return check1 ^ check2;
}

File: 2020/02.py
with open('inputs/02.in', 'r') as f:
    in_data = f.readlines()

def parse_row(row):
    parts = row.split()
    ranges = parts[0].split('-')
    return {
        'low': int(ranges[0]),
        'high': int(ranges[1]),
        'policy': parts[1].rstrip(':'),
        'password': parts[2]
    }

def part1(row):
    policy_range = list(range(row['low'], row['high']+1))
    return row['password'].count(row['policy']) in policy_range

def part2(row):
    indx1 = row['password'][row['low']-1] == row['policy']
    indx2 = row['password'][row['high']-1] == row['policy']
    return indx1 ^ indx2

parsed_rows = [parse_row(x) for x in in_data]
print('Part1:', sum(part1(x) for x in parsed_rows))
print('Part2:', sum(part2(x) for x in parsed_rows))
